var documenterSearchIndex = {"docs":
[{"location":"index.html#CompositeLD.jl-Documentation","page":"CompositeLD.jl Documentation","title":"CompositeLD.jl Documentation","text":"","category":"section"},{"location":"index.html#Overview","page":"CompositeLD.jl Documentation","title":"Overview","text":"","category":"section"},{"location":"index.html","page":"CompositeLD.jl Documentation","title":"CompositeLD.jl Documentation","text":"CompositeLD is a package using SnpArrays to perform some usefull calculations related LD. In particular :","category":"page"},{"location":"index.html","page":"CompositeLD.jl Documentation","title":"CompositeLD.jl Documentation","text":"Calculate LD between two SNPs (given position or id in PLINK files)\nGet the LD matrix of a group of SNPs\nGet Strongly correlated SNPs to a given group of SNPs\nClumping given SNPs in a prioritized order.","category":"page"},{"location":"index.html#Example","page":"CompositeLD.jl Documentation","title":"Example","text":"","category":"section"},{"location":"index.html","page":"CompositeLD.jl Documentation","title":"CompositeLD.jl Documentation","text":"using SnpArrays\nusing CompositeLD\n\ndata = SnpData(SnpArrays.datadir(datapath))\n\nld = ld_r2(\"rs1234\", \"rs654321\", data)\n\nld_mat = getLDmat(data, [(1, 459876), (1, 58735), (2, 97654)])\n\nstrong = getStrongLD(data, [\"rs1234\", \"rs5678\"])\n\nkept_v_b = clump(data, [(1, 459876), (1, 58735), (2, 97654)])","category":"page"},{"location":"index.html#Contents","page":"CompositeLD.jl Documentation","title":"Contents","text":"","category":"section"},{"location":"index.html","page":"CompositeLD.jl Documentation","title":"CompositeLD.jl Documentation","text":"","category":"page"},{"location":"index.html#Functions","page":"CompositeLD.jl Documentation","title":"Functions","text":"","category":"section"},{"location":"index.html","page":"CompositeLD.jl Documentation","title":"CompositeLD.jl Documentation","text":"formatSnpData!\n\nld_r2\n\ngetLDmat\n\ngetStrongLD\n\nclump\n\ntclump","category":"page"},{"location":"index.html#CompositeLD.formatSnpData!","page":"CompositeLD.jl Documentation","title":"CompositeLD.formatSnpData!","text":"format Genotype information contained in SnpData for snp sorted search based on chromosome and position (snpid). Adds a column chrpos (chr::Int8, pos::Int) in snpinfo and sorts snpinfo by chrpos (snpid). Adds a column containing original indices.     returns nothing\n\nExamples\n\nref = SnpData(datadir(\"some/data\"))\n\nformatSnpData!(ref, :chr_pos)\n\nformatSnpData!(ref, :snpid)\n\nNote this function does not take into account the user might want to write SnpData in bed, bim, fam format.      The changes done by this function can and will cause problems if writing plink files after calling formatSnpData!.\n\n\n\n\n\n","category":"function"},{"location":"index.html#CompositeLD.ld_r2","page":"CompositeLD.jl Documentation","title":"CompositeLD.ld_r2","text":"LD r² composite of pair of SNPs given genotype matrix and indices of snps in matrix\n\nExamples :\n\nref = SnpData(datadir(\"some/data\"))\n\nld_r2([0x00, 0x01, 0x03], [0x00, 0x00, 0x03])\n\nld_r2(1, 2, ref.snparray)\n\nformatSnpData!(ref, :snpid)\n\nld_r2(\"rs123\", \"rs456\", ref, formated = true)\n\nld_r2((1, 234), (1, 567), ref)\n\nIf formatSnpData has already been called on good snp info type (:chr_pos or :snpid), formated = true option does not verify or modify formating. See formatSnpData!. \nreturns a Float64\n\n\n\n\n\n","category":"function"},{"location":"index.html#CompositeLD.getLDmat","page":"CompositeLD.jl Documentation","title":"CompositeLD.getLDmat","text":"LD r² composite matrix of n SNPs from index in given SnpArray\n\nExamples :\n\nref = SnpData(datadir(\"some/data\"))\n\nM::Matrix{Float64} = getLDmat([2, 4, 8, 16], ref.snparray)\n\ngetLDmat([(1, 123), (1, 456), (1, 789)], ref)\n\nformatSnpData!(ref, :snpid)\n\nM::Matrix{Float64} = getLDmat([\"rs123\", \"rs456\", \"rs789\"], ref, formated = true)\n\nIf formatSnpData has already been called on good snp info type (:chr_pos or :snpid), formated = true option does not verify or modify formating. See formatSnpData!. \n\n\n\n\n\nGet correlation Matrix for specifies snps (tuple of the form (chr, pos)      given reference genotype SnpData.\n\n\n\n\n\n","category":"function"},{"location":"index.html#CompositeLD.getStrongLD","page":"CompositeLD.jl Documentation","title":"CompositeLD.getStrongLD","text":"Get strongly correlated variants from list.\n\narguments : \n\nref&#95;genotypes::SnpData : reference genotypes (formated by formatSnpData! or not) snps::Union{AbstractVector{<:Tuple{integer, Integer}}, AbstractVector{<:AbstractString}} : snps for which to search proxis\n\noptions : \n\nformated::Bool : indicates if ref SnpData is already formated according to :chr_pos (chr, pos) or :snpid (id as string)\nwindow::Integer : maximal distance from snip for ld calculations\nr2&#95;tresh::AbstractFloat : minimal r² for 2 snps to be considered strongly correlated.\n\nExamples\n\nref = SnpData(datadir(\"some/data\"))\n\nl::Vector{String} = getStrongLD([(1, 123), (1, 456), (1, 789)], ref, window = 250000)\n\nformatSnpData!(ref, :snpid)\n\nl::Vector{String} = getStrongLD([\"rs123\", \"rs456\", \"rs789\"], ref, formated = true, r2_tresh = 0.9)\n\nIf formatSnpData has already been called on good snp info type (:chr_pos or :snpid), formated = true option does not verify or modify formating. See formatSnpData!. \n\n\n\n\n\n\n","category":"function"},{"location":"index.html#CompositeLD.clump","page":"CompositeLD.jl Documentation","title":"CompositeLD.clump","text":"Implementation of the clumping algorithm prioritising first snps in given Vector and formated Genotypes SnpData (see formatSnpData!)     returns a vector of booean indication if each given snp is kept\n\narguments :\n\nref&#95;genotypes::SnpData : reference genotypes (formated by formatSnpData! or not) snps::Union{AbstractVector{<:Tuple{integer, Integer}}, AbstractVector{<:AbstractString}} : snps to clump\n\noptions : \n\nformated::Bool : indicates if ref SnpData is already formated according to :chr_pos (chr, pos) or :snpid (id as string)\nr2&#95;tresh::AbstractFloat : minimal r² for 2 snps to be considered strongly correlated.\n\nExamples :\n\njulia> ref = SnpData(datadir(\"some/data\"));\n\njulia> kept_v_b::Vector{Bool} = clump([(1, 123), (1, 456), (1, 789)], ref)\n3-element Vector{Int64}:\n 1\n 0\n 1\n\njulia> formatSnpData!(ref, :snpid);\n\njulia> kept_v_b::Vector{Bool} = clump([\"rs123\", \"rs456\", \"rs789\"], ref, formated = true)\n3-element Vector{Int64}:\n 1\n 0\n 1\n\nIf formatSnpData has already been called on good snp info type (:chr_pos or :snpid), formated = true option does not verify or modify formating. See formatSnpData!.\n\n\n\n\n\n","category":"function"},{"location":"index.html#CompositeLD.tclump","page":"CompositeLD.jl Documentation","title":"CompositeLD.tclump","text":"threaded implementation of the clumping algorithm prioritising first snps in given Vector and formated Genotypes SnpData (see formatSnpData!)     returns a vector of booean indication if each given snp is kept\n\narguments :\n\nref&#95;genotypes::SnpData : reference genotypes (formated by formatSnpData! or not) snps::Union{AbstractVector{<:Tuple{integer, Integer}}, AbstractVector{<:AbstractString}} : snps to clump\n\noptions : \n\nformated::Bool : indicates if ref SnpData is already formated according to :chr_pos (chr, pos) or :snpid (id as string)\nr2&#95;tresh::AbstractFloat : minimal r² for 2 snps to be considered strongly correlated.\n\nExamples :\n\nref = SnpData(datadir(\"some/data\"))\n\nkept_v_b::Vector{Bool} = tclump([(1, 123), (1, 456), (1, 789)], ref)\n\nformatSnpData!(ref, :snpid)\n\nkept_v_b::Vector{Bool} = tclump([\"rs123\", \"rs456\", \"rs789\"], ref, formated = true)\n\nIf formatSnpData has already been called on good snp info type (:chr_pos or :snpid), formated = true option does not verify or modify formating. See formatSnpData!.\n\n\n\n\n\n","category":"function"},{"location":"index.html#Index","page":"CompositeLD.jl Documentation","title":"Index","text":"","category":"section"},{"location":"index.html","page":"CompositeLD.jl Documentation","title":"CompositeLD.jl Documentation","text":"","category":"page"}]
}
